### ISSUE TITLE
`Promise.join` does not call callback if no promises provided

### ISSUE THREAD

**Issue by overlookmotel** [petkaantonov/bluebird#1218]
1) What version of bluebird is the issue happening on?

v2.10.2, v3.4.1

2) What platform and version? (For example Node.js 0.12 or Google Chrome 32)

Node v6.2.1, OS X 10.9.5

3) Did this issue happen with earlier version of bluebird?

Yes, consistent across v2.x and v3.x

---

(Originally raised as a side note in https://github.com/petkaantonov/bluebird/issues/1153#issuecomment-230173094)

The callback doesn't run if no promises are provided to `Promise.join()`.

```js
Promise.join( function() {
    console.log( 'callback!' );
} );
console.log('next sync statement');
```

outputs:

```
next sync statement
```

It would be pretty pointless for someone to call `Promise.join()` with no promises to join! But still you'd expect the callback to be called or an error to be thrown.

I suppose someone *might* run into this with:

```js
function foo( array ) {
    var after = function() { console.log( 'callback!' ); };
    return Promise.join.apply(Promise, array.concat( [ after ] ));
}
```

If `array` may or may not be empty, the callback being called most of the time, but skipped when the array is empty would be an unexpected result.

Of course, `Promise.all()` would be more sensible to use in this case, but you never know...

**Comment by benjamingr** [comment]
I think we should reject with a TypeError. This would be a semver major change.

**Comment by overlookmotel** [comment]
Should it also throw a TypeError if no parameters are provided at all? i.e. `Promise.join()` or `Promise.join.apply(Promise, [])`

I don't think this is covered by #1219.

**Comment by overlookmotel** [comment]
Oops. I just saw you made that change already.

**Comment by benjamingr** [comment]
@overlookmotel I did the change and reverted it since there is a test that checks that it passes on empty input. I think the test is from the "join works like `.all` but without the array" days. 

I think we can change that in 4.x - @petkaantonov ?

**Comment by overlookmotel** [comment]
@benjamingr Alternatively, don't throw a TypeError if no promises are provided, and just call the callback with an empty array?

**Comment by benjamingr** [comment]
I'll create a new branch with the alternative behavior so Petka can choose which he likes best.

**Comment by overlookmotel** [comment]
OK cool. Thank you. By the way, what kind of timescale are you thinking of for releasing v4.x?

**Comment by overlookmotel** [comment]
Not throwing a TypeError (the alternative behavior) wouldn't be a semver major change, would it?

**Comment by petkaantonov** [comment]
this is pretty gross misuse of Promise.join, definitely want to throw error here

**Comment by benjamingr** [comment]
Ok, going to change the behavior to throw a typeerror if a ufnction is not passed to `join`.

**Comment by benjamingr** [comment]
PTAL - I've removed the `join` code that doesn't take a function as an argument and fixed all the tests related to it.

**Comment by overlookmotel** [comment]
This would be a semver major change too, right? The docs state "In 1.x and 0.x Promise.join used to be a Promise.all that took the values in as arguments instead of an array. This behavior has been deprecated but is still supported partially - when the last argument is an immediate function value the new semantics will apply".

Personally, I quite like the old-style semantics of:

```js
Promise.join( p1, p2, p3 )
    .then( function(arr) {
        ...
    });
```

The `.then()` reminds me that the callback will be called async, and also I find it easier to read.

**Comment by benjamingr** [comment]
Right, this would be a semver major change, I'll push a doc update too.

**Comment by benjamingr** [comment]
You can just do:

```js
Promise.all([p1, p2, p3]).then(arr => {

});
```

The mental overhead isn't worth it. The reason join with the last argument is there is because it's faster.

**Comment by overlookmotel** [comment]
@benjamingr Fair enough. I didn't realise `.all()` would be as fast as `.join()` if you don't use the last arg callback style.
